### ISSUE TITLE
Consider sass-extract and sass-extract-js

### ISSUE THREAD

**Issue by penx** [penx/govuk-react#76]
Can we use govuk-frontend sass to ensure that our Glamorous components update when there are updates to the css lib?

https://medium.com/styled-components/getting-sassy-with-sass-styled-theme-9a375cfb78e8

**Comment by penx** [comment]
We would potentially aim to do this with https://github.com/alphagov/govuk-frontend/ once it is ready for use

**Comment by penx** [comment]
https://designnotes.blog.gov.uk/2018/02/19/developing-new-typography-and-spacing-for-gov-uk-frontend/

**Comment by marksy** [comment]
Yep, i have the sass-extract-js working in a branch. It was the wrong syntax on the example and also missing a dependency (sass-extract-loader). The sass-loader generates one large object so not sure what the best way to approach this would be... import just the button sass for our button component etc?

this will still generate a huge object because each component in govuk-frontend imports their global/common.scss, which in turn imports a whole bunch of stuff; settings tools, helpers, core, objects, overrides.

**Comment by igloosi** [comment]
hi, we're redoing this, to import only stuff it needs

**Comment by penx** [comment]
Blocked by https://github.com/alphagov/govuk-frontend/issues/636

**Comment by penx** [comment]
>  i have the sass-extract-js working in a branch

@marksy which branch?

**Comment by penx** [comment]
nm, think it's this one

https://github.com/UKHomeOffice/govuk-react/tree/feature/sassExtract

**Comment by penx** [comment]
I made a script that converts govuk-frontend in to emotion, it lives here https://github.com/penx/govuk-frontend-emotion

it's published on npm and there's example usage here https://github.com/penx/govuk-frontend-emotion-example

it's still a work in progress.. but it works

<img width="795" alt="screen shot 2018-08-17 at 20 36 42" src="https://user-images.githubusercontent.com/1027024/44285483-74cc3b80-a25d-11e8-993a-b228c32d98ad.png">

**Comment by marksy** [comment]
Hi yes that branch. Hope it is okay.

**Comment by penx** [comment]
we could probably extend this technique to make govuk-react, to some degree, work directly from the govuk-frontend styles

**Comment by penx** [comment]
we would still need to manually create React components using the appropriate DOM element that makes use of these styles, there may be a way to use the existing nunjucks or html templates to power this but probably a lot of work

**Comment by marksy** [comment]
You could probably stick it in a render function though

**Comment by penx** [comment]
Interesting interaction between govuk-frontend-toolkit and emotion here:

https://github.com/cds-snc/ircc-rescheduler/blob/master/src/components/forms/Button.js#L89

As linked from 
https://github.com/UKHomeOffice/govuk-react/issues/435#issuecomment-433373688

**Comment by pcraig3** [comment]
Hi there,

Yeah, the idea was to take keep the default button styles separated from the app-specific styling as a nod towards portability. In theory, the `govuk_button` rules could live somewhere else entirely (although then you would have to be pretty familiar with them to know what removing an override rule would do). 

Those styles never were used outside of that repo though, so there was never any problem of having to update a general set of rules and worrying about others downstream. 

It might be an evolutionary cul-de-sac, but it was super flexible for our project and let us pass styles around the app pretty easily (for example, our [`visuallyhidden` styles](https://github.com/cds-snc/ircc-rescheduler/blob/d2e3f888bcf7c8e3b78cff52dc5602dbc64fddb7/src/styles.js#L115-L129), which we could add to [any other component](https://github.com/cds-snc/ircc-rescheduler/search?q=visuallyhidden&unscoped_q=visuallyhidden) where convenient).

**Comment by penx** [comment]
Related https://github.com/facebook/create-react-app/issues/5687

**Comment by penx** [comment]
I'm not sure how we would deal with CSS such as this:

```scss
  .js-enabled {
    .govuk-header__menu-button {
      display: block;
      @include mq ($from: desktop) {
        display: none;
      }
    }
```

https://github.com/alphagov/govuk-frontend/blob/dd31d1d863be57d66ea285bbe8b6470dc84a1990/src/components/header/_header.scss#L189-L195

**Comment by penx** [comment]
Regarding js-enabled, there are 2 approaches:

1. have some js that manually adds this to the body (or similar) element and have a *.module.css that uses `:global(.js-enabled)` with the equivalent content of the *.css file.

2. use React to detect js-enabled via componentDidMount, perhaps using the context api to ensure it can be optimised to only run once.

This also highlights an important point about why we can't use govuk-frontend directly when doing React work, and there is need for a React component library that does more than just templating:

React development encourages you to not modify the DOM directly:

> React is unaware of changes made to the DOM outside of React. It determines updates based on its own internal representation, and if the same DOM nodes are manipulated by another library, React gets confused and has no way to recover.
https://reactjs.org/docs/integrating-with-other-libraries.html

(also see https://reactjs.org/docs/refs-and-the-dom.html)

govuk-frontend is dependent on scripts such as:

```js
  if ($toggleButton && $target) {
    this.toggleClass($target, 'govuk-header__navigation--open')
    this.toggleClass($toggleButton, 'govuk-header__menu-button--open')
```
https://github.com/alphagov/govuk-frontend/blob/master/src/components/header/header.js

Such scripts need to be rewritten to be handled via React rather than allowing direct DOM manipulation. In addition, the means a component like the Header has internal state (open or closed, and potentially js enabled or disabled). In React, it's good to [allow stateful components to be able to act as both controlled and uncontrolled](https://reactjs.org/docs/uncontrolled-components.html).

**Comment by marksy** [comment]
`.js-enabled` doesn't have to go on the body though. Can't it just be applied to the component (or it's parent)?

**Comment by penx** [comment]
@marksy yep :-) 

https://github.com/penx/govuk-frontend-react/commit/38a2a70d8162114f045dd6d5075f9a23f5903a37

**Comment by penx** [comment]
@stevesims @marksy @Loque- I am pretty convinced option 2 (via https://github.com/penx/govuk-frontend-react) is the way to go here.

I'm keen to know if anyone has any reservations on this approach on this so they can be considered/discussed before I spend any significant time heading in that direction :-)
